Fixes duplicate queries from MultilangstatusHelper

smz opened the pull request:
#### Description

several methods from MultilangstatusHelper are called multiple times in the life of a Joomla! process.
This PR optimize this by locally caching the results
#### Testing procedure
- Use a multilingual site
- Verify that everything is "_business as usual_" (a.k.a: _SNAFU_... :smile: ) with and without this PR, both in the front-end and in the back-end.
- Turn on Joomla Debug in global configuration and verify that several duplicated queries are now gone.

Even better: apply this together with #7137 and follow testing instruction there. If OK, you can give "@test OK" to both PRs
#### Additional comments

Thanks to "Viper" and "Robert G Mears" that have brought duplicate queries to my attention.
_This might be a first in a series..._

smz committed:
Fixes duplicate queries from getSitelangs()

smz commented:
I'm really unsure if this may have nasty side effects in the backend if/when you publish/unpublish languages... I should have tested that... :flushed:

smz commented:
... so far I can't see any issues generated by this...

smz committed:
CS

dgt41 commented on administrator/components/com_languages/helpers/multilangstatus.php:
```diff
@@ -91,7 +98,8 @@ public static function getSitelangs()
 		$db->setQuery($query);
 
-		return $db->loadObjectList('element');
+		$sitelangs = $db->loadObjectList('element');
```
Sergio can you wrap this to a try catch?

smz replied:
Dimitris, you mean to catch DB errors, like it is done in the new "Handle DB errors..." series of PRs?

In case, what should be, in your opinion, the returned value?

smz commented:
Dimitris, you mean to catch DB errors, like is it done in the new "*Handle DB errors...*" series of PRs?

smz committed:
Better CS (I guess...)

smz committed:
try/catch in DB query

smz commented:
@dgt41 Is that what you meant?
I guess throwing 500 is OK here...

smz commented:
Should we reserve the same *treatment* to all queries in this class?

dgt41 commented:
@smz Yes! Sorry for the late response…

smz commented:
:+1:  :smile:

smz committed:
Same treatment to more methods

I applied the same treatment to other methods that I *know* are called
more than once

smz commented:
Took the bull by its horns and applied the same treatment to other methods that I know are called
more than once...

Testing instructions do not change.

Even better: apply this together with #7137 and follow testing instruction there. If OK, you can give "@test OK" to both PRs

smz committed:
OUCH!

wilsonge commented on administrator/components/com_languages/helpers/multilangstatus.php:
```diff
@@ -81,17 +71,31 @@ public static function getContentlangs()
+			}
+			catch (RuntimeException $e)
+			{
+				throw new Exception($e->getMessage(), 500);
```
Is there any reason why we have to catch these exceptions and rethrow them and not just let them bubble up?

smz replied:
TBH, I just followed @dgt41 advice.
I've seen this is done in several new PRs, so I thought it was a good thing to do...
Better get rid of those?

mbabker replied:
Catching and rethrowing is actually worse than letting exceptions go uncaught.  It screws up the debug information contained in the Exception in that case.

If you're gonna catch the exception here, then you need to decide how to handle it.  If you're just going to catch and rethrow a new exception, you're better off letting it go uncaught and letting something higher in the stack catch it.

Typically, if the error condition isn't fatal to operation and things can go on with a "false" return, you can mock a return or something similar (in this case set `$sitelangs` to an empty object).  If it is a fatal error, then you should be working toward gracefully showing an error page.

smz replied:
OK: I'll remeve those try/catch, then!
Thanks Michael...

smz replied:
@mbabker and @wilsonge do you have 5 minute to come on Skype? I'm sergiomanzi there...

wilsonge replied:
Added ya

wilsonge commented on administrator/components/com_languages/helpers/multilangstatus.php:
```diff
@@ -101,18 +105,32 @@ public static function getSitelangs()
+			catch (RuntimeException $e)
+			{
+				throw new Exception($e->getMessage(), 500);
+			}
```
Ditto here

smz commented:
Forget about #7137: it was a *HUGE* mistake

smz commented:
*The reports of #7137's death have been greatly exaggerated.*

After all it can be done...

smz committed:
Remove try/catch

smz commented:
In Joomla Issues this results as "Pending" ( The Travis CI build is in progress), but actually Travis did his job already...

zero-24 commented:
@smz 

We have known issues on syncing travis results. If it passes here all is good :smile:

smz commented:
@zero-24 Thanks!

infograf768 commented:
I have not tested, but please do not delete the comments for each function.

smz commented:
The comments I'm removing are repetition of what is indicated just a couple of lines above in the docbloc... Do you really want me to reinstate them?

infograf768 commented:
No need, you are right.
